keps/sig-api-machinery/4358-custom-resource-field-selectors/README.md

```diff
@@ -0,0 +1,1150 @@
+| Pod                                       | status.nominatedNodeName           | string           | status.nominatedNodeName       |                                                               |
+| Secret                                    | type                               | enum             | type                           | string(enum)                                                  |
+
+[^1]: This field selector is commented with: "This is a bug, but we need to support it for backward compatibility.".
+
+Note that:
+
+- Resources always provide a value for a selectable field. Notice that in the
+  above table, *all* field selectors for optional fields provide a fallback if the
+  field is omitted.
+- Field selectors only exist for strings, enums, integers and booleans.
+- Field selectors for name and namespace are provided automatically.
+- Field selectors that do not match the path name of the actual field do exist,
+  for historical reasons, but the best practice is for field selectors to match
+  the exact path name of the field.
+
+### Plan
+
+Because `generic.StoreOptions` consumes selectors from built-in and CRDs in the
+same way, we would extend the `GetAttrs` of CRDs to offer fields for selection.
+https://github.com/jpbetz/kubernetes/tree/crd-object-filters contains a working
+prototype.
+
+We would leverage the fieldPath validation supported added for CRD Validation
+Rules to validate paths.
+
+We would leverage the jsonpath library in the same way as used by
+additionalPrinterColumns to retrieve values from custom resources using
+jsonpaths.
+
+The majority of the remaining work will be to update the API to add the
+field and to validate it.
+
+#### API
+
+```go
+type CustomResourceDefinitionVersion struct {
+  
+  // selectableFields specifies paths to fields that may be used as field selectors.
+  // A maximum of 8 selectable fields are allowed.
+	// See https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors
+	SelectableFields []SelectableField `json:"selectableFields,omitempty" protobuf:"bytes,9,rep,name=selectableFields"`
+}
+
+  // SelectableField specifies the JSON path of a field that may be used with field selectors.
+type SelectableField struct {
+	// jsonPath is a simple JSON path which is evaluated against each custom resource to produce a
+  // field selector value.
+  // Only JSON paths without the array notation are allowed.
+  // Must point to a valid field of type string, boolean or integer. Strings with enum values
+  // and strings with formats are allowed.
+  // Must not point to metdata fields.
+	JSONPath string `json:"jsonPath" protobuf:"bytes,1,opt,name=jsonPath"`
+
+  // valueIfAbsent provides a value for field selection if the value referred to
+  // by jsonPath is absent.
+  // valueIfAbsent is required unless all parts of jsonPath are required fields, or have defaults.
```

Comment:
is "" absent or present but zero value?

Reply:
default="" is present but zero valued.

The --field-selector implementation has a strong requirement that each resource in a list offer a string value for each selectableField. 

To meet this requirement, built-ins always provide a value for each selectableField. Some default to "", or "false", one backfills with a value from another field...  "Conversions/Notes" in [the table](https://github.com/kubernetes/enhancements/blob/1504d0aaba219a18a0b3c2f9c17940cf0a7e82a0/keps/sig-api-machinery/4358-custom-resource-field-selectors/README.md#background-field-selection-as-it-exists-before-this-enhancement) has more details.

For CRDs, A default value will be required for any selectableFields that can be absent from a resource (e.g. the field is optional and has no default)